standard/expressions.md

```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+    : Interpolated_Regular_String_Start Interpolated_Regular_String_Mid?
+      ('{' regular_interpolation '}' Interpolated_Regular_String_Mid?)* Interpolated_Regular_String_End
+    ;
+
+regular_interpolation
+    : expression (',' interpolation_minimum_width)? Regular_Interpolation_Format?
+    ;
+
+interpolation_minimum_width
+    : constant_expression
+    ;
+
+Interpolated_Regular_String_Start
+    : '$"'
+    ;
+
+// the following three lexical rules are context sensitive, see details below
+Interpolated_Regular_String_Mid
+    : Interpolated_Regular_String_Element+
+    ;
+
+Regular_Interpolation_Format
+    : ':' Interpolated_Regular_String_Element+
+    ;
+
+Interpolated_Regular_String_End
+    : '"'
+    ;
+
+fragment Interpolated_Regular_String_Element
+    : Interpolated_Regular_String_Character
+    | Simple_Escape_Sequence
+    | Hexadecimal_Escape_Sequence
+    | Unicode_Escape_Sequence
+    | Open_Brace_Escape_Sequence
+    | Close_Brace_Escape_Sequence
+    ;
+
+fragment Interpolated_Regular_String_Character
+    // Any character except " (U+0022), \\ (U+005C), { (U+007B), } (U+007D), and New_Line_Character
+    : ~["\\{}\u000D\u000A\u0085\u2028\u2029]
+    ;
+
+// interpolated verbatim string expressions
+
+interpolated_verbatim_string_expression
+    : Interpolated_Verbatim_String_Start Interpolated_Verbatim_String_Mid?
+      ('{' verbatim_interpolation '}' Interpolated_Verbatim_String_Mid?)* Interpolated_Verbatim_String_End
+    ;
+
+verbatim_interpolation
+    : expression (',' interpolation_minimum_width)? Verbatim_Interpolation_Format?
+    ;
+
+Interpolated_Verbatim_String_Start
+    : '$@"'
+    ;
+
+// the following three lexical rules are context sensitive, see details below
+Interpolated_Verbatim_String_Mid
+    : Interpolated_Verbatim_String_Element+
+    ;
+
+Verbatim_Interpolation_Format
+    : ':' Interpolated_Verbatim_String_Element+
+    ;
+
+Interpolated_Verbatim_String_End
+    : '"'
+    ;
+
+fragment Interpolated_Verbatim_String_Element
+    : Interpolated_Verbatim_String_Character
+    | Quote_Escape_Sequence
+    | Open_Brace_Escape_Sequence
+    | Close_Brace_Escape_Sequence
+    ;
+
+fragment Interpolated_Verbatim_String_Character
+    // Any character except " (U+0022), { (U+007B) and } (U+007D)
+    : ~["{}]
+    ;
+
+// lexical fragments used by both regular and verbatim interpolated strings
+
+fragment Open_Brace_Escape_Sequence
+    : '{{'
+    ;
+
+fragment Close_Brace_Escape_Sequence
+    : '}}'
+    ;
+```
+
+Six of the lexical rules defined above are *context sensitive* as follows:
+
+| Rule | Contextual Requirements |
+| :--- | :---------------------- |
+| *Interpolated_Regular_String_Mid* | Only recognised after an *Interpolated_Regular_String_Start*, between any following interpolations, and before the corresponding *Interpolated_Regular_String_End*. |
+| *Regular_Interpolation_Format* | Only recognised immediately before the closing delimiter (`}`) of a *regular_interpolation*. |
```

Comment:
The tricky part here is the format can contain just about anything, including parentheses etc. So the rule matches if there is a colon, some text, and then a following '}' (which is not part of the token itself) *and* the whole token is at the outmost level.

What we're not making clear here, because its an implementation thing, is that in order to implement the context sensitivity you need to track *all* bracket kinds.

The question is whether the text here is a correct/sufficient description so that implementors can produce a correct implementation. I'll elaborate on the TG2 list…

Reply:
I like the alternative you suggested to the committee.  Adding it here for discussion:

```suggestion
| *Regular_Interpolation_Format* | Only recognised within a *regular_interpolation*, when the starting colon (`:`) is not nested within any kind of bracket (parentheses/braces/square), and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*. |
```

Reply:
Agreed with the suggestion without `, and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*`.